### ISSUE TITLE
feat: Impulse based (instead of force player position) indicator slam + more accurate slam damage - [merged]

### ISSUE THREAD

**Issue by MaxwellJung** [Overwatch-1-Emulator/ow1-emulator#1559]
In GitLab by @hikiru on Nov 6, 2023, 15:34

_Merges indicator_rework -> staging_

- Indicator slam now uses impulses to allow for object collision, and it no longer messes up camera control
- Fixed ground slam damage
- Added a couple lines of code to uppercut so its gravity changes don't interfere with indicator

To-do:
- See if it's possible to make redirection feel and function smoother
- Limit x-axis deviation when redirecting to prevent going around large objects and walls
- Fix slams where the player gets stuck and doesn't drop

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 15:38

added 2 commits

<ul><li>ed412458 - 1 commit from branch <code>staging</code></li><li>d6107801 - Reworked indicator slam with other minor changes</li></ul>

[Compare with previous version](/MaxwellJung/ow1_emulator/-/merge_requests/30/diffs?diff_id=836881305&start_sha=82af3f1e983321c2078c59f478bc9e3ce3d4f6f8)

**Comment by MaxwellJung** [comment]
```suggestion:-21+0
```

You can safely delete this code (it's prserved in version history)

**Comment by MaxwellJung** [comment]
```suggestion:-0+0
        eventPlayer.applyImpulse(directionTowards(eventPlayer.getPosition(), eventPlayer.indicator_slam_position), OW1_DOOMFIST_INDICATOR_SEISMIC_SLAM_SPEED, Relativity.TO_WORLD, Impulse.CANCEL_CONTRARY_MOTION)
```

Replace magic number with constant

**Comment by MaxwellJung** [comment]
```suggestion:-0+0
        eventPlayer.applyImpulse(-eventPlayer.getVelocity(), eventPlayer.getSpeed() - OW1_DOOMFIST_INDICATOR_SEISMIC_SLAM_SPEED, Relativity.TO_WORLD, Impulse.INCORPORATE_CONTRARY_MOTION) # somewhat prevents doom from speeding up after getting caught on objects
```

Another magic number

**Comment by MaxwellJung** [comment]
```suggestion:-0+0
    cancelMomentum(eventPlayer)
```

Replace with macrofunction from `utilities/macro_functions.opy` for readability

**Comment by MaxwellJung** [comment]
Why is this conditional on _only_ ground slam? Why not just remove if check?

**Comment by MaxwellJung** [comment]
```suggestion:-2+0
    if (eventPlayer.isUsingAbility1()): return
    eventPlayer.setGravity(50)
```

Shorten to single line (I just like it better asthetically)

**Comment by MaxwellJung** [comment]
```suggestion:-1+0
    if (eventPlayer.isUsingAbility1()): return
```

Same comment as before

**Comment by MaxwellJung** [comment]
```suggestion:-4+0
#Conditions for when to stop an indicator slam and drop the player
#!define indicatorCanContinue() \
    (distance(eventPlayer.getPosition(), eventPlayer.indicator_slam_position) > 2) \
and (eventPlayer.isUsingAbility1()) \
and (eventPlayer.getSpeed() > 0) \
```

Indent for clarity

**Comment by MaxwellJung** [comment]
Overall logic looks good; just need some changes to match code styling guidelines.

**Comment by MaxwellJung** [comment]
requested review from @MaxwellJung

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 15:56

Commented on [src/ow1/heroes/doomfist/slam.opy line 216](https://github.com/Overwatch-1-Emulator/ow1-emulator/compare/ed412458f3fd228d83651944512784e1b1745291..bad5a0b7a6c60a738d243b46b960eb8a940230f5#diff-c3ea8c2a418a3dd19cdbba970a815f69R216)



The wait only fixes ground slam damage and not indicator

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 15:58

Commented on [src/ow1/heroes/doomfist/slam.opy line 178](https://github.com/Overwatch-1-Emulator/ow1-emulator/compare/ed412458f3fd228d83651944512784e1b1745291..d61078011685196ebec83ccb5c5ac191424a2366#diff-c3ea8c2a418a3dd19cdbba970a815f69R178)



changed this line in [version 3 of the diff](/MaxwellJung/ow1_emulator/-/merge_requests/30/diffs?diff_id=836888593&start_sha=d61078011685196ebec83ccb5c5ac191424a2366#7fca664a602119d818c549db0367f308a59f70ed_178_157)

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 15:58

Commented on [src/ow1/heroes/doomfist/slam.opy line 192](https://github.com/Overwatch-1-Emulator/ow1-emulator/compare/ed412458f3fd228d83651944512784e1b1745291..d61078011685196ebec83ccb5c5ac191424a2366#diff-c3ea8c2a418a3dd19cdbba970a815f69R192)



changed this line in [version 3 of the diff](/MaxwellJung/ow1_emulator/-/merge_requests/30/diffs?diff_id=836888593&start_sha=d61078011685196ebec83ccb5c5ac191424a2366#7fca664a602119d818c549db0367f308a59f70ed_192_170)

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 15:58

Commented on [src/ow1/heroes/doomfist/slam.opy line 198](https://github.com/Overwatch-1-Emulator/ow1-emulator/compare/ed412458f3fd228d83651944512784e1b1745291..d61078011685196ebec83ccb5c5ac191424a2366#diff-c3ea8c2a418a3dd19cdbba970a815f69R198)



changed this line in [version 3 of the diff](/MaxwellJung/ow1_emulator/-/merge_requests/30/diffs?diff_id=836888593&start_sha=d61078011685196ebec83ccb5c5ac191424a2366#7fca664a602119d818c549db0367f308a59f70ed_198_176)

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 15:58

Commented on [src/ow1/heroes/doomfist/uppercut.opy line 60](https://github.com/Overwatch-1-Emulator/ow1-emulator/compare/ed412458f3fd228d83651944512784e1b1745291..d61078011685196ebec83ccb5c5ac191424a2366#diff-d8ad29b9e68fef0c5556a38aac618edfR60)



changed this line in [version 3 of the diff](/MaxwellJung/ow1_emulator/-/merge_requests/30/diffs?diff_id=836888593&start_sha=d61078011685196ebec83ccb5c5ac191424a2366#96388e4435957f563ef7af1679afd75971556a99_60_58)

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 15:58

added 1 commit

<ul><li>e7d297e3 - Apply 6 suggestion(s) to 2 file(s)</li></ul>

[Compare with previous version](/MaxwellJung/ow1_emulator/-/merge_requests/30/diffs?diff_id=836888593&start_sha=d61078011685196ebec83ccb5c5ac191424a2366)

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 15:59

Commented on [src/ow1/heroes/doomfist/slam.opy line 172](https://github.com/Overwatch-1-Emulator/ow1-emulator/compare/ed412458f3fd228d83651944512784e1b1745291..e7d297e3a192394138acfdfb079473b9b2e18bdc#diff-c3ea8c2a418a3dd19cdbba970a815f69R172)



changed this line in [version 4 of the diff](/MaxwellJung/ow1_emulator/-/merge_requests/30/diffs?diff_id=836889048&start_sha=e7d297e3a192394138acfdfb079473b9b2e18bdc#7fca664a602119d818c549db0367f308a59f70ed_172_172)

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 15:59

added 1 commit

<ul><li>bad5a0b7 - Apply 1 suggestion(s) to 1 file(s)</li></ul>

[Compare with previous version](/MaxwellJung/ow1_emulator/-/merge_requests/30/diffs?diff_id=836889048&start_sha=e7d297e3a192394138acfdfb079473b9b2e18bdc)

**Comment by MaxwellJung** [comment]
If the intent of that code is to delay the damage ramp up by 1 tick, I think it applies to both ground and indicator slam and the if statement can be removed.

**Comment by MaxwellJung** [comment]
```suggestion:-2+0
    waitUntil(not eventPlayer.isOnGround(), Math.INFINITY)
    if eventPlayer.slam_to_use == SlamType.GROUND: # Makes ground slam damage 49    
        wait()
```

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 16:05

Commented on [src/ow1/heroes/doomfist/slam.opy line 216](https://github.com/Overwatch-1-Emulator/ow1-emulator/compare/ed412458f3fd228d83651944512784e1b1745291..bad5a0b7a6c60a738d243b46b960eb8a940230f5#diff-c3ea8c2a418a3dd19cdbba970a815f69R216)



Max range indicator damage gives 65 damage as is most of the time, sometimes I get 67 but any changes that reduce slam damage makes indicator always 63 damage and I can't get it up to 65. Indicator damage is very finicky for some reason

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 16:09

Commented on [src/ow1/heroes/doomfist/slam.opy line 216](https://github.com/Overwatch-1-Emulator/ow1-emulator/compare/ed412458f3fd228d83651944512784e1b1745291..bad5a0b7a6c60a738d243b46b960eb8a940230f5#diff-c3ea8c2a418a3dd19cdbba970a815f69R216)



changed this line in [version 5 of the diff](/MaxwellJung/ow1_emulator/-/merge_requests/30/diffs?diff_id=836893764&start_sha=bad5a0b7a6c60a738d243b46b960eb8a940230f5#7fca664a602119d818c549db0367f308a59f70ed_216_216)

**Comment by MaxwellJung** [comment]
In GitLab by @hikiru on Nov 6, 2023, 16:09

added 1 commit

<ul><li>993d04dd - Apply 1 suggestion(s) to 1 file(s)</li></ul>

[Compare with previous version](/MaxwellJung/ow1_emulator/-/merge_requests/30/diffs?diff_id=836893764&start_sha=bad5a0b7a6c60a738d243b46b960eb8a940230f5)

**Comment by MaxwellJung** [comment]
resolved all threads

**Comment by MaxwellJung** [comment]
marked this merge request as **ready**

**Comment by MaxwellJung** [comment]
approved this merge request

**Comment by MaxwellJung** [comment]
mentioned in commit 216e6018abfa27febce9f579bb49a8cf41db49a2

**Comment by MaxwellJung** [comment]
In GitLab by @ghost1 on Nov 7, 2023, 11:03

:tada: This MR is included in version 1.10.0 :tada:

The release is available on [GitLab release](https://gitlab.com/MaxwellJung/ow1_emulator/-/releases/v1.10.0).

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package: :rocket:

**Comment by MaxwellJung** [comment]
fixes #367 #297 #116 #104
